### PR TITLE
openshift: add vault-manager deployment template

### PR DESCRIPTION
**Pull request by TGPSKI**
This PR adds an Openshift deployment template for vault-manager.

### CODE REVIEW

**Comment by TGPSKI** [comment]
[test]

**Comment by TGPSKI** [comment]
[test]

File: Dockerfile [hunk 1]
```diff
@@ -1,8 +1,8 @@
-FROM registry.access.redhat.com/ubi8/go-toolset:1.13.15 as builder
+FROM registry.access.redhat.com/ubi8/go-toolset:1.15.13 as builder
 COPY . .
 RUN make gobuild
 
-FROM registry.access.redhat.com/ubi8-minimal:8.2
+FROM registry.access.redhat.com/ubi8-minimal:8.4
```
Comment (maorfr) [inline Dockerfile:5]: all non-template related changes really

**Review by maorfr — CHANGES_REQUESTED** [review]
some nits above, but mainly:
1. i'm not seeing the dry-run option.
2. if i'm not mistaken, the created pod will run the vault-manager command once and exit. we need to run this in a loop with sleep time in between.

File: Dockerfile [hunk 2]
```diff
@@ -1,8 +1,8 @@
-FROM registry.access.redhat.com/ubi8/go-toolset:1.13.15 as builder
+FROM registry.access.redhat.com/ubi8/go-toolset:1.15.13 as builder
```
[the remark this answers is not in the document]
Comment (TGPSKI) [reply Dockerfile:1]: My first thought was that the changes to CentOS fork strategy in 2020 had a side effect on the availability of this repository, so this was just a quick hack to try and get tests to pass before I continued. 

I agree with you 100%

File: openshift/vault-manager.yaml [hunk 3]
```diff
@@ -0,0 +1,113 @@
...
+  value: 800Mi
+- name: MEMORY_LIMIT
+  value: 1Gi
+- name: CPU_REQUESTS
+  value: 500m
+- name: CPU_LIMIT
+  value: 1000m
+# These values are meant to be overridden by the
+# saas-herder parametrization
+- name: GITHUB_USERS
```
[the remark this answers is not in the document]
Comment (TGPSKI) [reply openshift/vault-manager.yaml:99]: I included some lines from the source I borrowed from, focus was on getting the pr check to pass before I went back over these values with a fine comb. PR is still in draft state =]

[the remark this answers is not in the document]
Comment (maorfr) [reply openshift/vault-manager.yaml:99]: apologies!

[the remark this answers is not in the document]
Comment (TGPSKI) [reply openshift/vault-manager.yaml:99]: You are all good! I appreciate your willingness to get in here and help me along, you're great.

File: openshift/vault-manager.template.yaml [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+        - image: ${IMAGE}:${IMAGE_TAG}
+          imagePullPolicy: Always
+          name: vault-manager
+          env:
+            - name: GRAPHQL_QUERY_FILE
+              value: ${GRAPHQL_QUERY_FILE}
+            - name: GRAPHQL_SERVER_URL
+              valueFrom:
+                secretKeyRef:
+                  key: gql.base_url
```
Comment (TGPSKI) [inline openshift/vault-manager.template.yaml:37]: Can the secret injector put two different secrets in the same environment context with a key?

```
gql.base_url
gql.username
gql.password
vault.auth_type
vault.role_id
vault.secret_id
```

If not, then this works:

```
```
base_url
username
password
auth_type
role_id
secret_id
```
```

File: openshift/vault-manager.template.yaml [hunk 5]
```diff
@@ -0,0 +1,95 @@
...
+          app: vault-manager
+      spec:
+        containers:
+        - image: ${IMAGE}:${IMAGE_TAG}
+          imagePullPolicy: Always
+          name: vault-manager
+          command: ["/bin/sh"]
+          args: 
+          - "-c"
+          - "while true; do /vault-manager; sleep 15m; done;"
```
Comment (TGPSKI) [inline openshift/vault-manager.template.yaml:34]: Added periodic command

Comment (maorfr) [reply openshift/vault-manager.template.yaml:34]: can we make the sleep time a parameter?

File: openshift/vault-manager.template.yaml [hunk 6]
```diff
@@ -0,0 +1,94 @@
...
+          env:
+          - name: GRAPHQL_QUERY_FILE
+            value: ${GRAPHQL_QUERY_FILE}
+          - name: GRAPHQL_SERVER_URL
+            valueFrom:
+              secretKeyRef:
+                key: gql.base_url
+                name: ${GRAPHQL_SERVER_URL}
```
Comment (maorfr) [inline openshift/vault-manager.template.yaml:42]: `name` is the Secret's name, we can probably just hard code it to be `vault-manager`? or something like VAULT_MANAGER_SECRET_NAME if you prefer that the secret name is parameterized?

File: openshift/vault-manager.template.yaml [hunk 7]
```diff
@@ -0,0 +1,94 @@
...
+  displayName: vault-manager image
+  description: vault-manager docker image. Defaults to quay.io/app-sre/vault-manager
+- name: IMAGE_TAG
+  value: latest
+  displayName: vault-manager version
+  description: vault-manager version which defaults to latest
+- name: MEMORY_REQUESTS
+  value: 200Mi
+- name: MEMORY_LIMIT
+  value: 300Mi
```
Comment (maorfr) [inline openshift/vault-manager.template.yaml:88]: did you profile this?

File: openshift/vault-manager.template.yaml (hunk 4 above)
Comment (maorfr) [reply openshift/vault-manager.template.yaml]: https://gitlab.cee.redhat.com/service/app-interface#manage-secrets-via-app-interface-openshiftnamespace-1yml-using-vault

**Review by maorfr — COMMENTED** [review]
this is starting to come together! don't forget to add the dry-run option.

File: openshift/vault-manager.template.yaml [hunk 8]
```diff
@@ -0,0 +1,98 @@
...
+          app: vault-manager
+      spec:
+        containers:
+        - image: ${IMAGE}:${IMAGE_TAG}
+          imagePullPolicy: Always
+          name: vault-manager
+          command: ["/bin/sh"]
+          args: 
+          - "-c"
+          - "while true; do /vault-manager {{ if ${DRY_RUN} == 'True' }}--dry-run{{ end }}; sleep ${RECONCILE_SLEEP_TIME}; done;"
```
Comment (TGPSKI) [inline openshift/vault-manager.template.yaml:34]: Following Helm conditional conventions here. I haven't seen a conditional in a few other templates I checked out, so hoping this approach works.

File: openshift/vault-manager.template.yaml [hunk 9]
```diff
@@ -0,0 +1,98 @@
...
+  value: quay.io/app-sre/vault-manager
+  displayName: vault-manager image
+  description: vault-manager docker image. Defaults to quay.io/app-sre/vault-manager
+- name: IMAGE_TAG
+  value: latest
+  displayName: vault-manager version
+  description: vault-manager version which defaults to latest
+- name: MEMORY_REQUESTS
+  value: 200Mi
+- name: MEMORY_LIMIT
```
Comment (TGPSKI) [inline openshift/vault-manager.template.yaml:87]: I bumped this up for now. docker stats on the running container didn't show much in terms of memory or cpu usage. There are some deeper profiling methods I can get into.

What's the thought on dialing these experimentally via kube container metrics on a staging deployment?

File: openshift/vault-manager.template.yaml (hunk 8 above)
Comment (TGPSKI) [reply openshift/vault-manager.template.yaml:34]: Nevermind, `oc process` does not support conditionals. Moving to a two template approach to support --dry-run.

Comment (TGPSKI) [reply openshift/vault-manager.template.yaml]: An alternative solution (probably better than two templates) is to set a DRY_RUN parameter with a default of : ""

```
- "while true; do /vault-manager $DRY_RUN; sleep ${RECONCILE_SLEEP_TIME}; done;"
...
parameters:
- name: DRY_RUN
  value: ""
  description: enables the dry-run cli flag when value is set to "--dry-run"
```

File: openshift/vault-manager.template.yaml (hunk 9 above)
Comment (maorfr) [reply openshift/vault-manager.template.yaml:87]: i think it's ok. you can also look at https://blog.golang.org/pprof.

File: openshift/vault-manager.template.yaml [hunk 10]
```diff
@@ -0,0 +1,99 @@
...
+  value: 250m
+- name: CPU_LIMIT
+  value: 500m
+- name: REPLICAS
+  value: '1'
+- name: RECONCILE_SLEEP_TIME
+  value: '15m'
+- name: DRY_RUN
+  description: enables dry run cli flag when value is set to "--dry-run"
+  value: ''
```
Comment (maorfr) [inline openshift/vault-manager.template.yaml:99]: missing a new line

Comment (maorfr) [reply openshift/vault-manager.template.yaml:99]: you can also probably pass true/false to the flag?

File: openshift/vault-manager.template.yaml (hunk 8 above)
Comment (maorfr) [reply openshift/vault-manager.template.yaml]: you can also probably pass true/false to the flag? i also think it's `-dry-run` with a single dash?

**Review by maorfr — APPROVED** [review]
despite the small nits, let's get this going! we can iterate in stage (just make sure dry-run is true).

**Pull request merged**